src/compiler.rs

```diff
@@ -153,4 +194,86 @@ impl Compiler {
 
         rbp.span_offset
     }
+
+    /// Enter a scope, e.g., a block or a closure
+    pub fn enter_scope(&mut self) {
+        self.scope.push(Frame {
+            frame_type: FrameType::Scope,
+            vars: HashMap::new(),
+        });
+    }
+
+    /// Exit a scope, e.g., when reaching the end of a block or a closure
+    ///
+    /// When exiting a scope, all overlays (and corresponding light scope frames) are removed along
+    /// with the removed scope.
+    pub fn exit_scope(&mut self) {
+        match self
+            .scope
+            .iter()
+            .rposition(|frame| frame.frame_type == FrameType::Scope)
+        {
+            None => panic!("no scope frame to exit"),
+            Some(0) => panic!("can't exit the top-most scope frame"),
+            Some(pos) => self.scope.truncate(pos),
+        }
+    }
+
+    /// Get a span of a node
+    pub fn get_span(&self, node_id: NodeId) -> Span {
+        *self.spans.get(node_id.0).expect("node doesn't have span")
+    }
+
+    /// Get the source contents under the span
+    pub fn get_span_contents(&self, span: Span) -> &[u8] {
+        self.source
+            .get(span.start..span.end)
+            .expect("tried getting invalid span contents")
+    }
+
+    /// Fetch last scope frame that can be written to
+    pub fn last_frame_mut(&mut self) -> &mut Frame {
+        for frame in self.scope.iter_mut().rev() {
+            if frame.frame_type == FrameType::Scope || frame.frame_type == FrameType::Light {
+                return frame;
+            }
+        }
+
+        panic!("No valid scope frame");
+    }
+
+    pub fn add_variable(&mut self, name: Vec<u8>, var_id: NodeId) {
+        self.last_frame_mut().vars.insert(name, var_id);
+    }
+
+    pub fn find_variable(&self, name: &[u8]) -> Option<NodeId> {
+        for frame in self.scope.iter().rev() {
+            if let Some(var_id) = frame.vars.get(name) {
+                return Some(*var_id);
+            }
+        }
+
+        None
+    }
+
+    pub fn get_variable(&self, var_id: NodeId) -> &AstNode {
+        self.ast_nodes
+            .get(var_id.0)
+            .expect("tried getting invalid variable AST node")
+    }
+
+    /// Flatten the scope frames into only one frame
```

Comment:
What's the flattening into one frame do? I mean like what's a good place it will get used in the future?

Reply:
Nothing, you haven't seen anything, it's never been there 😆. I was just playing around and forgot to delete it. It would give you a view of all current definitions in the Compiler, but I don't have a use case for it (in Nushell, we do the same thing before sending env vars to an external).